File: examples/wikipedia_concept_fetcher.py
#!/usr/bin/env python3
"""
Wikipedia Concept Fetcher (Python)

Demonstrates how to fetch canonical concept data from Wikipedia and Wikidata
for use in MECS Learning Course Builder.

Requirements:
    pip install requests

Usage:
    python examples/wikipedia_concept_fetcher.py "Arithmetic"
    python examples/wikipedia_concept_fetcher.py "Q11205"
"""

import sys
import json
import requests
from datetime import datetime
from typing import Dict, List, Optional, Any


def search_wikidata(query: str) -> List[Dict[str, Any]]:
    """Search Wikidata for a concept by name."""
    url = "https://www.wikidata.org/w/api.php"
    params = {
        "action": "wbsearchentities",
        "search": query,
        "language": "en",
        "format": "json",
    }

    print(f"🔍 Searching Wikidata for: '{query}'")

    response = requests.get(url, params=params)
    response.raise_for_status()
    data = response.json()

    return data.get("search", [])


def get_wikidata_entity(wikidata_id: str) -> Dict[str, Any]:
    """Get detailed entity data from Wikidata."""
    url = f"https://www.wikidata.org/wiki/Special:EntityData/{wikidata_id}.json"

    print(f"📥 Fetching Wikidata entity: {wikidata_id}")

    response = requests.get(url)
    response.raise_for_status()
    data = response.json()

    return data["entities"][wikidata_id]


def get_wikipedia_summary(title: str) -> Dict[str, Any]:
    """Get Wikipedia page summary."""
    url = f"https://en.wikipedia.org/api/rest_v1/page/summary/{title}"

    print(f"📄 Fetching Wikipedia summary: {title}")

    response = requests.get(url)
    response.raise_for_status()

    return response.json()


def get_wikipedia_categories(title: str) -> List[str]:
    """Get Wikipedia page categories."""
    url = "https://en.wikipedia.org/w/api.php"
    params = {
        "action": "query",
        "titles": title,
        "prop": "categories",
        "format": "json",
    }

    response = requests.get(url, params=params)
    response.raise_for_status()
    data = response.json()

    pages = data["query"]["pages"]
    page = list(pages.values())[0]

    if "categories" in page:
        return [cat["title"].replace("Category:", "") for cat in page["categories"]]

    return []


def get_related_concepts(entity: Dict[str, Any]) -> List[Dict[str, str]]:
    """Get related concepts from Wikidata."""
    related = []
    claims = entity.get("claims", {})

    # Get "has part" (P527)
    has_part = claims.get("P527", [])
    for claim in has_part:
        value = claim.get("mainsnak", {}).get("datavalue", {}).get("value", {})
        if "id" in value:
            related.append({"wikidataId": value["id"], "relationship": "has-part"})

    # Get "part of" (P361)
    part_of = claims.get("P361", [])
    for claim in part_of:
        value = claim.get("mainsnak", {}).get("datavalue", {}).get("value", {})
        if "id" in value:
            related.append({"wikidataId": value["id"], "relationship": "part-of"})

    return related


def create_canonical_concept(wikidata_id: str) -> Dict[str, Any]:
    """Create a canonical concept reference from Wikidata + Wikipedia data."""

    # 1. Fetch Wikidata entity
    entity = get_wikidata_entity(wikidata_id)

    if not entity:
        raise ValueError(f"Wikidata entity {wikidata_id} not found")

    # 2. Extract basic info
    title = entity.get("labels", {}).get("en", {}).get("value")
    description = entity.get("descriptions", {}).get("en", {}).get("value")
    aliases = [a["value"] for a in entity.get("aliases", {}).get("en", [])]
    wikipedia_title = entity.get("sitelinks", {}).get("enwiki", {}).get("title")

    if not title:
        raise ValueError("Missing English label")

    if not wikipedia_title:
        print("⚠️  No English Wikipedia article found")

    # 3. Fetch Wikipedia data (if available)
    wikipedia_summary = None
    categories = []
    wikipedia_url = None

    if wikipedia_title:
        try:
            wikipedia_summary = get_wikipedia_summary(wikipedia_title)
            categories = get_wikipedia_categories(wikipedia_title)
            wikipedia_url = wikipedia_summary["content_urls"]["desktop"]["page"]
        except Exception as e:
            print(f"⚠️  Could not fetch Wikipedia data: {e}")

    # 4. Get related concepts
    related_concepts = get_related_concepts(entity)

    # 5. Get instance of (P31)
    instance_of = None
    claims_p31 = entity.get("claims", {}).get("P31", [])
    if claims_p31:
        instance_of = (
            claims_p31[0].get("mainsnak", {}).get("datavalue", {}).get("value", {}).get("id")
        )

    # 6. Create canonical concept
    canonical_concept = {
        "conceptId": wikidata_id,
        "title": title,
        "sourceUrl": wikipedia_url or f"https://www.wikidata.org/wiki/{wikidata_id}",
        "sourceType": "wikipedia" if wikipedia_url else "wikidata",
        "metadata": {
            "wikidataId": wikidata_id,
            "wikidataUrl": f"https://www.wikidata.org/wiki/{wikidata_id}",
            "description": description,
            "aliases": aliases,
            "instanceOf": instance_of,
            "extract": wikipedia_summary.get("extract") if wikipedia_summary else None,
            "thumbnail": wikipedia_summary.get("thumbnail", {}).get("source")
            if wikipedia_summary
            else None,
            "categories": categories,
            "relatedConcepts": related_concepts,
            "lastVerified": datetime.utcnow().isoformat() + "Z",
        },
    }

    return canonical_concept


def main():
    """Main function."""
    if len(sys.argv) < 2:
        print(
            """
Usage:
  python examples/wikipedia_concept_fetcher.py <search-term>
  python examples/wikipedia_concept_fetcher.py <wikidata-id>

Examples:
  python examples/wikipedia_concept_fetcher.py "Arithmetic"
  python examples/wikipedia_concept_fetcher.py "Q11205"
  python examples/wikipedia_concept_fetcher.py "Photosynthesis"
  python examples/wikipedia_concept_fetcher.py "Linear Algebra"
        """
        )
        sys.exit(1)

    input_query = sys.argv[1]

    try:
        wikidata_id = None

        # Check if input is already a Wikidata ID
        if input_query.startswith("Q") and input_query[1:].isdigit():
            wikidata_id = input_query
            print(f"✅ Using Wikidata ID: {wikidata_id}\n")
        else:
            # Search for the concept
            print(f"🔎 Searching for: '{input_query}'\n")
            results = search_wikidata(input_query)

            if not results:
                print(f"❌ No results found for '{input_query}'")
                sys.exit(1)

            print("📋 Search Results:\n")
            for i, result in enumerate(results[:5], 1):
                print(f"{i}. {result['label']} ({result['id']})")
                if result.get("description"):
                    print(f"   {result['description']}")
                print()

            # Use the first result
            wikidata_id = results[0]["id"]
            print(f"✅ Using top result: {results[0]['label']} ({wikidata_id})\n")

        # Fetch and create canonical concept
        print("📦 Creating canonical concept...\n")
        concept = create_canonical_concept(wikidata_id)

        # Display results
        print("\n" + "=" * 80)
        print("CANONICAL CONCEPT")
        print("=" * 80 + "\n")

        print(json.dumps(concept, indent=2))

        print("\n" + "=" * 80)
        print("MECS COURSE METADATA (Example)")
        print("=" * 80 + "\n")

        mecs_metadata = {
            "mecsVersion": "0.3.0",
            "type": "mecs:course",
            "id": f"{concept['conceptId'].lower()}-example",
            "title": f"Introduction to {concept['title']}",
            "metadata": {
                "canonicalConcept": {
                    "conceptId": concept["conceptId"],
                    "title": concept["title"],
                    "sourceUrl": concept["sourceUrl"],
                    "sourceType": concept["sourceType"],
                },
                "description": concept["metadata"]["description"],
                "learningType": "concept",
                "academicLevel": "undergrad-lower",
                "proficiencyLevel": "understand",
            },
        }

        print(json.dumps(mecs_metadata, indent=2))

        print("\n✅ Done!\n")

    except Exception as e:
        print(f"\n❌ Error: {e}")
        import traceback

        traceback.print_exc()
        sys.exit(1)


if __name__ == "__main__":
    main()

File: examples/wikipedia-concept-fetcher.js
/**
 * Wikipedia Concept Fetcher
 *
 * Demonstrates how to fetch canonical concept data from Wikipedia and Wikidata
 * for use in MECS Learning Course Builder.
 *
 * Usage:
 *   node examples/wikipedia-concept-fetcher.js "Arithmetic"
 *   node examples/wikipedia-concept-fetcher.js "Q11205"
 */

// Simple fetch for Node.js (works in modern Node versions)
const fetch = (...args) => import('node-fetch').then(({default: fetch}) => fetch(...args));

/**
 * Search Wikidata for a concept by name
 */
async function searchWikidata(query) {
  const url = `https://www.wikidata.org/w/api.php?action=wbsearchentities&search=${encodeURIComponent(query)}&language=en&format=json&origin=*`;

  console.log(`🔍 Searching Wikidata for: "${query}"`);

  const response = await fetch(url);
  const data = await response.json();

  return data.search;
}

/**
 * Get detailed entity data from Wikidata
 */
async function getWikidataEntity(wikidataId) {
  const url = `https://www.wikidata.org/wiki/Special:EntityData/${wikidataId}.json`;

  console.log(`📥 Fetching Wikidata entity: ${wikidataId}`);

  const response = await fetch(url);
  const data = await response.json();

  return data.entities[wikidataId];
}

/**
 * Get Wikipedia page summary
 */
async function getWikipediaSummary(title) {
  const url = `https://en.wikipedia.org/api/rest_v1/page/summary/${encodeURIComponent(title)}`;

  console.log(`📄 Fetching Wikipedia summary: ${title}`);

  const response = await fetch(url);

  if (!response.ok) {
    throw new Error(`Wikipedia API error: ${response.status}`);
  }

  return await response.json();
}

/**
 * Get Wikipedia page categories
 */
async function getWikipediaCategories(title) {
  const url = `https://en.wikipedia.org/w/api.php?action=query&titles=${encodeURIComponent(title)}&prop=categories&format=json&origin=*`;

  const response = await fetch(url);
  const data = await response.json();

  const pages = Object.values(data.query.pages)[0];

  if (pages.categories) {
    return pages.categories.map(c => c.title.replace('Category:', ''));
  }

  return [];
}

/**
 * Get related concepts from Wikidata
 */
function getRelatedConcepts(entity) {
  const related = [];

  // Get "has part" (P527)
  const hasPart = entity.claims?.P527 || [];
  for (const claim of hasPart) {
    const value = claim.mainsnak?.datavalue?.value;
    if (value?.id) {
      related.push({
        wikidataId: value.id,
        relationship: 'has-part'
      });
    }
  }

  // Get "part of" (P361)
  const partOf = entity.claims?.P361 || [];
  for (const claim of partOf) {
    const value = claim.mainsnak?.datavalue?.value;
    if (value?.id) {
      related.push({
        wikidataId: value.id,
        relationship: 'part-of'
      });
    }
  }

  return related;
}

/**
 * Create a canonical concept reference from Wikidata + Wikipedia data
 */
async function createCanonicalConcept(wikidataId) {
  // 1. Fetch Wikidata entity
  const entity = await getWikidataEntity(wikidataId);

  if (!entity) {
    throw new Error(`Wikidata entity ${wikidataId} not found`);
  }

  // 2. Extract basic info
  const title = entity.labels?.en?.value;
  const description = entity.descriptions?.en?.value;
  const aliases = entity.aliases?.en?.map(a => a.value) || [];
  const wikipediaTitle = entity.sitelinks?.enwiki?.title;

  if (!title) {
    throw new Error("Missing English label");
  }

  if (!wikipediaTitle) {
    console.warn("⚠️  No English Wikipedia article found");
  }

  // 3. Fetch Wikipedia data (if available)
  let wikipediaSummary = null;
  let categories = [];
  let wikipediaUrl = null;

  if (wikipediaTitle) {
    try {
      wikipediaSummary = await getWikipediaSummary(wikipediaTitle);
      categories = await getWikipediaCategories(wikipediaTitle);
      wikipediaUrl = wikipediaSummary.content_urls.desktop.page;
    } catch (error) {
      console.warn(`⚠️  Could not fetch Wikipedia data: ${error.message}`);
    }
  }

  // 4. Get related concepts
  const relatedConcepts = getRelatedConcepts(entity);

  // 5. Get instance of (P31)
  const instanceOf = entity.claims?.P31?.[0]?.mainsnak?.datavalue?.value?.id;

  // 6. Create canonical concept
  const canonicalConcept = {
    conceptId: wikidataId,
    title: title,
    sourceUrl: wikipediaUrl || `https://www.wikidata.org/wiki/${wikidataId}`,
    sourceType: wikipediaUrl ? "wikipedia" : "wikidata",
    metadata: {
      wikidataId: wikidataId,
      wikidataUrl: `https://www.wikidata.org/wiki/${wikidataId}`,
      description: description,
      aliases: aliases,
      instanceOf: instanceOf,
      extract: wikipediaSummary?.extract,
      thumbnail: wikipediaSummary?.thumbnail?.source,
      categories: categories,
      relatedConcepts: relatedConcepts,
      lastVerified: new Date().toISOString()
    }
  };

  return canonicalConcept;
}

/**
 * Main function
 */
async function main() {
  const args = process.argv.slice(2);

  if (args.length === 0) {
    console.log(`
Usage:
  node examples/wikipedia-concept-fetcher.js <search-term>
  node examples/wikipedia-concept-fetcher.js <wikidata-id>

Examples:
  node examples/wikipedia-concept-fetcher.js "Arithmetic"
  node examples/wikipedia-concept-fetcher.js "Q11205"
  node examples/wikipedia-concept-fetcher.js "Photosynthesis"
  node examples/wikipedia-concept-fetcher.js "Linear Algebra"
    `);
    process.exit(1);
  }

  const input = args[0];

  try {
    let wikidataId;

    // Check if input is already a Wikidata ID
    if (input.match(/^Q\d+$/)) {
      wikidataId = input;
      console.log(`✅ Using Wikidata ID: ${wikidataId}\n`);
    } else {
      // Search for the concept
      console.log(`🔎 Searching for: "${input}"\n`);
      const results = await searchWikidata(input);

      if (results.length === 0) {
        console.error(`❌ No results found for "${input}"`);
        process.exit(1);
      }

      console.log(`📋 Search Results:\n`);
      results.slice(0, 5).forEach((result, i) => {
        console.log(`${i + 1}. ${result.label} (${result.id})`);
        if (result.description) {
          console.log(`   ${result.description}`);
        }
        console.log();
      });

      // Use the first result
      wikidataId = results[0].id;
      console.log(`✅ Using top result: ${results[0].label} (${wikidataId})\n`);
    }

    // Fetch and create canonical concept
    console.log(`📦 Creating canonical concept...\n`);
    const concept = await createCanonicalConcept(wikidataId);

    // Display results
    console.log(`\n${'='.repeat(80)}`);
    console.log(`CANONICAL CONCEPT`);
    console.log(`${'='.repeat(80)}\n`);

    console.log(JSON.stringify(concept, null, 2));

    console.log(`\n${'='.repeat(80)}`);
    console.log(`MECS COURSE METADATA (Example)`);
    console.log(`${'='.repeat(80)}\n`);

    const mecsMetadata = {
      "mecsVersion": "0.3.0",
      "type": "mecs:course",
      "id": `${concept.conceptId.toLowerCase()}-example`,
      "title": `Introduction to ${concept.title}`,
      "metadata": {
        "canonicalConcept": {
          "conceptId": concept.conceptId,
          "title": concept.title,
          "sourceUrl": concept.sourceUrl,
          "sourceType": concept.sourceType
        },
        "description": concept.metadata.description,
        "learningType": "concept",
        "academicLevel": "undergrad-lower",
        "proficiencyLevel": "understand"
      }
    };

    console.log(JSON.stringify(mecsMetadata, null, 2));

    console.log(`\n✅ Done!\n`);

  } catch (error) {
    console.error(`\n❌ Error: ${error.message}`);
    console.error(error.stack);
    process.exit(1);
  }
}

// Run if called directly
if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}

// Also support CommonJS
if (typeof require !== 'undefined' && require.main === module) {
  main();
}
